core/src/main/scala/org/apache/spark/executor/ProcfsBasedSystems.scala

```diff
@@ -0,0 +1,268 @@
+      // https://docs.oracle.com/javase/9/docs/api/java/lang/ProcessHandle.html
+      val cmd = Array("bash", "-c", "echo $PPID")
+      val length = 10
+      val out: Array[Byte] = Array.fill[Byte](length)(0)
+      Runtime.getRuntime.exec(cmd).getInputStream.read(out)
+      val pid = Integer.parseInt(new String(out, "UTF-8").trim)
+      return pid;
+    }
+    catch {
+      case e: IOException => logDebug("IO Exception when trying to compute process tree." +
+        " As a result reporting of ProcessTree metrics is stopped")
+        isAvailable = false
+        return -1
+      case _ => logDebug("Some exception occurred when trying to compute process tree. " +
+        "As a result reporting of ProcessTree metrics is stopped")
+        isAvailable = false
+        return -1
+    }
+  }
+
+  private def computePageSize(): Unit = {
+    val cmd = Array("getconf", "PAGESIZE")
+    val out: Array[Byte] = Array.fill[Byte](10)(0)
+    Runtime.getRuntime.exec(cmd).getInputStream.read(out)
+    pageSize = Integer.parseInt(new String(out, "UTF-8").trim)
+  }
+
+  private def computeProcessTree(): Unit = {
+    if (!isAvailable) {
+      return
+    }
+    computePageSize
+    val queue: Queue[Int] = new Queue[Int]()
+    queue += pid
+    while( !queue.isEmpty ) {
+      val p = queue.dequeue()
+      val c = getChildPIds(p)
+      if(!c.isEmpty) {
+        queue ++= c
+        ptree += (p -> c.toSet)
+      }
+      else {
+        ptree += (p -> Set[Int]())
+      }
+    }
+  }
+
+  private def getChildPIds(pid: Int): ArrayBuffer[Int] = {
+    try {
+      val cmd = Array("pgrep", "-P", pid.toString)
+      val input = Runtime.getRuntime.exec(cmd).getInputStream
+      val childPidsInByte: mutable.ArrayBuffer[Byte] = new mutable.ArrayBuffer()
+      var d = input.read()
+      while (d != -1) {
+        childPidsInByte.append(d.asInstanceOf[Byte])
+        d = input.read()
+      }
+      input.close()
+      val childPids = new String(childPidsInByte.toArray, "UTF-8").split("\n")
+      val childPidsInInt: ArrayBuffer[Int] = new ArrayBuffer[Int]()
+      for (p <- childPids) {
+        if (p != "") {
+          childPidsInInt += Integer.parseInt(p)
+        }
+      }
+      childPidsInInt
+    } catch {
+      case e: IOException => logDebug("IO Exception when trying to compute process tree." +
+        " As a result reporting of ProcessTree metrics is stopped")
+        isAvailable = false
+        return new mutable.ArrayBuffer()
+      case _ => logDebug("Some exception occurred when trying to compute process tree." +
+        " As a result reporting of ProcessTree metrics is stopped")
+        isAvailable = false
+        return new mutable.ArrayBuffer()
+    }
+  }
+
+  /**
+   * Hadoop ProcfsBasedProcessTree class used regex and pattern matching to retrive the memory
+   * info. I tried that but found it not correct during tests, so I used normal string analysis
+   * instead. The computation of RSS and Vmem are based on proc(5):
```

Comment:
Can you expand on what you found was not correct during tests?  is that covered by the unit tests you've added?

and super nit: this is more of a regular implementation comment, not a docstring -- so `/*` not `/**`, and also maybe move it inside the method

Reply:
Yeah, the unit test cover this. I noticed that the it is hard to verify whether the regex match with the proc(5).